### PR TITLE
Test fixes for date-range using selection model

### PR DESCRIPTION
**Pull request by MathiasPaumgarten**
First batch of fixes.

### CODE REVIEW

File: src/lib/datepicker/calendar-body.spec.ts [hunk 1]
```diff
@@ -70,6 +74,7 @@ describe('MatCalendarBody', () => {
           .toBe(cellEls.length - 1, 'Expected remaining cells to be marked as deselected.');
     });
 
+    // tslint:disable-next-line:ban
```
Comment (MathiasPaumgarten) [inline src/lib/datepicker/calendar-body.spec.ts]: Remove.

File: src/lib/datepicker/year-view.ts [hunk 2]
```diff
@@ -27,11 +27,11 @@ import {
   EventEmitter,
   Inject,
   Input,
-  OnDestroy,
   Optional,
   Output,
   ViewChild,
   ViewEncapsulation,
+  OnDestroy,
```
Comment (MathiasPaumgarten) [inline src/lib/datepicker/year-view.ts]: unnecessary change.

File: src/lib/datepicker/calendar-body.spec.ts (hunk 1 above)
Comment (mmalerba) [reply src/lib/datepicker/calendar-body.spec.ts]: I assume this is not needed anymore?

File: src/lib/datepicker/calendar.ts [hunk 3]
```diff
@@ -24,6 +24,7 @@ import {
   SimpleChanges,
   ViewChild,
   ViewEncapsulation,
+  SkipSelf,
```
Comment (mmalerba) [inline src/lib/datepicker/calendar.ts]: It looks like this isn't used in this file

File: src/lib/datepicker/datepicker-input.ts [hunk 4]
```diff
@@ -134,20 +134,21 @@ export class MatDatepickerInput<D> implements ControlValueAccessor, OnDestroy, V
...
 
-    if (!this._dateAdapter.sameDate(value, this._selectionModel.getFirstSelectedDate())) {
+    if (!this._dateAdapter.sameDate(value, oldDate)) {
       this._selectionModel.setSelection(value);
     }
 
-    this._lastValueValid = !this._selectionModel || this._selectionModel.isValid();
+    this._lastValueValid = this._selectionModel.isValid();
 
     this._formatValue(this._selectionModel.getFirstSelectedDate());
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker-input.ts]: `getFirstSelectedDate()` --> `getSelection()`

File: src/lib/datepicker/datepicker-input.ts [hunk 5]
```diff
@@ -339,17 +340,14 @@ export class MatDatepickerInput<D> implements ControlValueAccessor, OnDestroy, V
   }
 
   _onInput(value: string) {
-    const date = this._dateAdapter.parse(value, this._dateFormats.parse.dateInput);
+    let date = this._dateAdapter.parse(value, this._dateFormats.parse.dateInput);
     const current = this._selectionModel.getFirstSelectedDate();
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker-input.ts]: `getFirstSelectedDate()` --> `getSelection()`

File: src/lib/datepicker/datepicker-input.ts [hunk 6]
```diff
@@ -380,7 +378,8 @@ export class MatDatepickerInput<D> implements ControlValueAccessor, OnDestroy, V
     }
 
     this._elementRef.nativeElement.value =
-        value ? this._dateAdapter.format(value, this._dateFormats.display.dateInput) : '';
+        value && this._getValidDateOrNull(value) ?
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker-input.ts]: not necessary to check `value &&` since `_getValidDateOrNull` accepts `any`

Comment (MathiasPaumgarten) [reply src/lib/datepicker/datepicker-input.ts]: True, but the compiler will complain in the follow up line in `this._dateAdapter.format(value...` since format doesn't accept null.

File: src/lib/datepicker/calendar-body.spec.ts [hunk 7]
```diff
@@ -117,22 +121,70 @@ describe('MatCalendarBody', () => {
...
   selectedValue = 4;
   labelMinRequiredCells = 3;
   numCols = 7;
 
+  constructor() {
+    const daysInMonth = daysInCurrentMonth();
+    const rows = this.createStackedNumbers(daysInMonth);
+
+    // We create rows based on the current month. This ensures that today is included.
+    this.rows = rows.map(r => r.map(this.createCell));
```
Comment (MathiasPaumgarten) [inline src/lib/datepicker/calendar-body.spec.ts]: I'm creating a month that covers as man days as this current month. I do this to ensure that the current calculation works. Today used to be passed in but now the calendar determines that itself. I just realized that I could simply do that by overriding the adapters `today()` method with a mock. 
I think I should revert this change.

File: src/lib/datepicker/calendar.ts [hunk 8]
```diff
@@ -236,7 +237,11 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
   /** A function used to filter which dates are selectable. */
   @Input() dateFilter: (date: D) => boolean;
 
-  /** Emits when the currently selected date changes. */
+  /**
+   * Emits when the currently selected date changes.
+   * @deprecated Listen to selectionModel valueChange.
+   * @breaking-change: 9.0.0
```
Comment (mmalerba) [inline src/lib/datepicker/calendar.ts]: nit: no `:` after `@breaking-change`

File: src/lib/datepicker/month-view.ts [hunk 9]
```diff
@@ -88,7 +88,7 @@ export class MatMonthView<D> implements AfterContentInit, OnDestroy {
   get selected(): D | null { return this._selectionModel.getFirstSelectedDate(); }
   set selected(value: D | null) {
     if (this._selectionModel instanceof MatSingleDateSelectionModel) {
-      this._selectionModel.setSelection(value);
+      this._selectionModel.add(value);
```
Comment (mmalerba) [inline src/lib/datepicker/month-view.ts]: Why `add` instead of `setSelection`? I think they both do the same in the end, `setSelection` makes more sense to me though

File: src/lib/core/datetime/date-selection.ts [hunk 10]
```diff
@@ -0,0 +1,150 @@
+/**
```
Comment (mmalerba) [inline src/lib/core/datetime/date-selection.ts]: why does this file show as new?

File: src/lib/datepicker/month-view.ts (hunk 9 above)
Comment (MathiasPaumgarten) [reply src/lib/datepicker/month-view.ts]: Not entirely. `setSelection` will not notify anyone that the value has changed. It changes it silently. I need other components to update when I set this.

File: src/lib/core/datetime/date-selection.ts (hunk 10 above)
Comment (MathiasPaumgarten) [reply src/lib/core/datetime/date-selection.ts]: I have no idea 😕 Some GIT merging issue?? Should I look into it or do we not bother as it will all be merged into one commit eventually.
